src/components/ThemeProductSection.tsx

```diff
@@ -0,0 +1,125 @@
+import Grid from "./Grid"
+import Card from "./Card"
+import Text from "./Text"
+import ProductImage from "./ProductImage"
+import ThemeNotFound from "./PresentTheme/ThemeNotFound"
+import Loading from "./PresentTheme/Loading"
+import theme from "@/styles/theme"
+import { Product } from "@/interfaces/Product"
+import useThemeProduct from "@/hooks/useThemeProduct"
+import useInfiniteScrolling from "@/hooks/useInfiniteScrolling"
+
+import { useState, useEffect, useCallback } from "react"
+import { useNavigate } from "react-router-dom"
+import { ROUTES } from "@/constants/routes"
+import getRoute from "@/functions/getRoute"
+import { useAuth } from "@/context/AuthContext"
+
+async function fetchThemeProducts(
+  themeId: string,
+  cursor: number | null,
+  limit = 12,
+) {
+  const base = import.meta.env.VITE_BASE_URL
+  const url = new URL(`/api/themes/${themeId}/products`, base)
+  if (cursor !== null) url.searchParams.append("cursor", String(cursor))
+  url.searchParams.append("limit", String(limit))
+
+  const res = await fetch(url.toString()).then(r => r.json())
+  return res.data
+}
+
+
+const ThemeProductSection = ({ themeId }: { themeId: string }) => {
+
+  const [products, setProducts] = useState<Product[]>([])
+  const [cursor, setCursor]     = useState<number | null>(null)
+  const [hasMore, setHasMore]   = useState(true)
+  const [isFetching, setIsFetching] = useState(false)
+  const [observerRef, setObserverRef] =
+    useState<null | HTMLDivElement>(null)
+
+  const {
+    list: initialList,
+    cursor: initialCursor,
+    hasMore: initialHasMore,
+    loading,
+    error,
+  } = useThemeProduct(themeId) 
+
+  useEffect(() => {
+    if (initialList.length) {
+      setProducts(initialList)
+      setCursor(initialCursor ?? null)
+      setHasMore(initialHasMore)
+    }
+  }, [initialList, initialCursor, initialHasMore])
+
+
+  const fetchMore = useCallback(async () => {
+    if (!hasMore || isFetching) return      
+    setIsFetching(true)
+    try {
+    const res = await fetchThemeProducts(themeId, cursor)
+    setProducts(prev => [...prev, ...res.list])
+    setCursor(res.cursor)
+    setHasMore(res.hasMoreList)
+}
+    finally {
+        setIsFetching(false)                 
+    }
+  }, [themeId, cursor, hasMore])
```

Comment:
[변경 요청] 페이지네이션으로 API를 받아오는 부분이 추상화 되지 않고 컴포넌트 코드에 작성되어 있어서 컴포넌트의 가독성이 떨어지는 것 같아요. product에 대한 도메인 의존성은 완전히 분리하지 않더라도, 아래처럼 페이지네이션 관리를 위한 상태들을 useThemeProduct으로 이동하는 것만으로도 개선될 것 같아요.

```tsx
  const {
    products,
    hasMore,
    isFetching,
    error,
    hasMore,
    loadMore,
  } = useThemeProducts();
```

```suggestion
  const [observerRef, setObserverRef] =
    useState<null | HTMLDivElement>(null)

  const {
    products,
    hasMore,
    loading,
    error,
    fetchMore
  } = useThemeProduct(themeId);
```

Reply:
넵 멘토님 이동하는 방향으로 수정하였습니다!